### DISCUSSION TITLE
Preloading / Caching

### DISCUSSION THREAD

**Discussion by lordcraymen** [google/model-viewer#4427] [Q&A] [answered]
On my site I will have several models the viewer can switch back and forth between. I wanted to have the models cross fade each other. I overlayed model viewer instances on top of each other, but this naive approach obviously leads to poor performance. I guess what I could to is take a screenshot of the current model and apply it in the loading or poster slot and fade it out when the new model is ready.

https://youtube.com/shorts/ZCudHEkdWGo?feature=share  This is a video of what the cross fading looks like.

I just wondered, does it make sense to preload models? I thought about having a zero pixel modelViewer instance invisible in the page, going through all the other not currently visible models and set the src every time a model is ready. There are other methods of using link prefetch, but I am using glb and gltf with external binaries an textures, so preloading just the gltf without any binary resources wouldn’t mean much…

Does anyone have an idea how to handle preloading and caching of multiple models?

**Comment by elalish** [comment]
We actually handle multiple MV elements very efficiently - that's part of our value prop. So long as your MV element is not visible, we'll stop doing anything with it. We already use a single shared renderer for all elements. I would recommend against creating and destroying MV elements a lot - we usually handle a carousel using three MV elements and on swipe you can move the last one up to the front and set the new src (you can use `loading="eager"` to prefetch).

**Reply by lordcraymen** [reply]
Thank you for the reply. I will check if the other model viewers are Indeed display:none or visibility:hidden, and if that makes a difference. It would be much nicer with multiple mv on top of each other as the fading element would also animate, almost like a morphing transition.

**Reply by lordcraymen** [reply]
As I expected display:none and visibility:hidden make a huge difference. When visibility:hidden is set and the same model is loaded e.g. 5 times in 5 instances the 3d performance drops drastically and does not change wether the elements are visible or not. When set to display none the performance is as expected. 

I need the visibility to work like hidden (except the reflow that happens with display none), can I add this with code? I’m thinking of creating a fork anyway, I just wonder if there is any logic that checks for display in the model viewer. Thanks for your support!

**Reply by elalish** [reply, marked as answer]
I'm not quite sure what you mean. As usual there are probably 100 different ways to solve this with CSS, but I'm not an expert. MV is not specifically watching the display property, but rather being informed that its element is not visible (removed from DOM, outside intersection observer, etc). If you need to control reflow, you could try swapping out the invisible MV element with an empty div of the same dimensions.

**Reply by elalish** [reply]
You can test whether we consider the element to be visible at any given time with https://modelviewer.dev/docs/index.html#entrydocs-loading-properties-modelIsVisible

**Reply by lordcraymen** [reply]
I see, I tested it and the „hack“ I can use is to move the mv out of the window with css when it is considered invisible by my code. It probably works with display:none because well it’s not rendered in the Dom and the intersection observer does not find it. Thanks for your pointers, it helped me a lot!
